db/system_keyspace.cc

```diff
@@ -2676,8 +2676,13 @@ service::topology_features system_keyspace::decode_topology_features_state(::sha
     for (auto& row : *rs) {
         raft::server_id host_id{row.get_as<utils::UUID>("host_id")};
         service::node_state nstate = service::node_state_from_string(row.get_as<sstring>("node_state"));
-        if (row.has("supported_features") && nstate == service::node_state::normal) {
-            ret.normal_supported_features.emplace(host_id, decode_features(deserialize_set_column(*topology(), row, "supported_features")));
+        if (row.has("supported_features")) {
+            auto features = decode_features(deserialize_set_column(*topology(), row, "supported_features"));
+            if (nstate == service::node_state::normal) {
+                ret.normal_supported_features.emplace(host_id, std::move(features));
+            } else if (nstate == service::node_state::none || nstate == service::node_state::replacing) {
```

Comment:
Now I see how it is used. We do not store features in the regular node maps, but have this structure instead. As far as I understand this is for early loading. But OTOH we need to only early load features for normal nodes. May be we can have a separate structure for early check, but also store features together with all other node's metadata as well.

Reply:
I agree 100%. Supported features used to be a field of `replica_state` before, but I moved them to a separate struct in https://github.com/scylladb/scylladb/pull/14955 because they were needed to be loaded early. However, I agree that both having this struct and `supported_features` field in `replica_state` as you propose is more convenient. I refactored the code in this way in v2.